compiler/src/lib.rs

```diff
@@ -65,16 +73,81 @@ pub fn compile(
                 chunk_id,
                 is_script: false,
             };
-            let name = &env.fqn;
-            let signature_idx = cp.insert_string(name.to_string());
-            bytecode.emit_constant_ref(signature_idx);
-            compile_chunk(chunk, chunk_id, &env.variables, &mut bytecode, ctx, &mut cp);
+            let name = env.fqn.clone();
+            compile_chunk(
+                name,
+                chunk,
+                chunk_id,
+                ctx,
+                &mut bytecode,
+                &mut cp,
+                line_provider,
+            );
         }
     }
 
     write(writer, &bytecode, &cp)
 }
 
+fn compile_chunk(
+    name: Name,
+    chunk: &Chunk,
+    id: SourceId,
+    ctx: EmitterContext,
+    bytecode: &mut Bytecode,
+    cp: &mut ConstantPool,
+    line_provider: Option<&dyn SourceLineProvider>,
+) {
+    // emit the function's name
+    let signature_idx = cp.insert_string(name);
+    bytecode.emit_constant_ref(signature_idx);
+
+    // emits chunk's code attribute
+    let segments = compile_chunk_code(chunk, id, bytecode, ctx, cp);
+
+    bytecode.emit_byte(line_provider.map_or(0, |_| 1));
+
+    if let Some(line_provider) = line_provider {
+        let Some(content_id) = ctx.engine.get_original_content(id) else { return };
+        compile_line_mapping_attribute(segments, content_id, bytecode, line_provider);
+    }
+}
+
+fn compile_line_mapping_attribute(
+    positions: Vec<(usize, u32)>,
+    content_id: ContentId,
+    bytecode: &mut Bytecode,
+    line_provider: &dyn SourceLineProvider,
+) {
+    bytecode.emit_byte(MAPPINGS_ATTRIBUTE);
+    let mut mappings: Vec<(usize, u32)> = Vec::new();
+
+    let Some(((mut last_pos, mut last_ip), positions)) = positions.split_first() else {
+        bytecode.emit_u32(0);
+        return
+    };
+    let mut last_line = usize::MAX;
```

Comment:
The initial position is never inserted.

Reply:
it is expected to be masked out when there are many positions, i added a case to effectively append the first pos' line to the first instruction (#0) if `mappings` is empty